### PR TITLE
Refactor agent controller

### PR DESCRIPTION
**Pull request by pablochacin**
# Description

Refactor the code around the AgentController to improve the separation of concerns and facilitate testing by separating the logic that handles the management of a set of pods (PodController) from the logic that interacts with one agent (PodVisitor).

Also simplifies the API around the generation of the commands used for interacting with the pod agent.

# Checklist:

- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works.   
- [X] I have run linter locally (`make lint`) and all checks pass.
- [X] I have run tests locally (`make test`) and all tests pass.
- [X] I have run relevant integration test locally (`make integration-xxx` for affected packages)
- [X] I have run relevant e2e test locally (`make e2e-xxx` for `disruptors`, or `cluster` related changes)
- [ ] Any dependent changes have been merged and published in downstream modules<br>

### CODE REVIEW

File: pkg/disruptors/controller_test.go [hunk 1]
```diff
@@ -2,63 +2,102 @@ package disruptors
...
+}
+
+func visitCommands() PodVisitCommand {
+	return fakeCommand{
+		exec:    []string{"command"},
+		cleanup: []string{"cleanup"},
+	}
+}
+
+func Test_VisitPod(t *testing.T) {
```
Comment (roobre) [inline pkg/disruptors/controller_test.go:36]: With the current refactor, this is now testing the `PodAgentVisitor` specifically, not any other implementation of `PodVisitor`. Should we rename the test to represent this?

File: pkg/disruptors/controller_test.go [hunk 2]
```diff
@@ -68,125 +107,108 @@ func Test_InjectAgent(t *testing.T) {
...
-	return v.cmds, v.err
+func (f fakePodVisitor) Visit(_ context.Context, _ corev1.Pod) error {
+	time.Sleep(f.delay)
+	return f.err
 }
 
-func Test_VisitPod(t *testing.T) {
+var errFailed = errors.New("failed")
+
+func Test_AgentControler(t *testing.T) {
```
Comment (roobre) [inline pkg/disruptors/controller_test.go:159]: Small typo
```suggestion
func Test_AgentController(t *testing.T) {
```

File: pkg/disruptors/pod_test.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func Test_NewPodDisruptor(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		title       string
+		name        string
+		namespace   string
```
Comment (roobre) [inline pkg/disruptors/pod_test.go:23]: What are `name` and `namespace` used for? They don't seem to be used in the test body.

**Review by roobre — COMMENTED** [review]
I'm _really_ liking these changes, the code feels much more natural now. Awesome work!

I left some comments, mostly related to docstrings and a few regarding naming. LMK what you think!

File: pkg/disruptors/controller.go [hunk 4]
```diff
@@ -14,40 +14,58 @@ import (
...
 )
 
-// PodVisitor defines the interface for visiting Pods
-type PodVisitor interface {
-	// Visit returns the VisitComands for visiting the Pod
-	Visit(pod corev1.Pod) (VisitCommands, error)
+// PodVisitCommand define the commands used for visiting a Pod
+type PodVisitCommand interface {
+	// Commands defines the command to be executed, and optionally a cleanup command
+	Commands(corev1.Pod) ([]string, []string, error)
```
[the remark this answers is not in the document]
Comment (pablochacin) [reply pkg/disruptors/controller.go:20]: No strong reason, only deduce the boiler plate. Can revert this change.

File: pkg/disruptors/controller.go [hunk 5]
```diff
@@ -72,135 +90,98 @@ func (c *agentController) InjectDisruptorAgent(ctx context.Context) error {
...
+	// if the context is cancelled, don't report error (we assume the caller is reporting this error)
+	if err != nil && !errors.Is(err, context.Canceled) {
+		return fmt.Errorf("failed command execution for pod %q: %w \n%s", pod.Name, err, string(stderr))
+	}
+
+	return nil
+}
+
+// NewAgentController creates a new controller for a fleet of  pods
+func NewAgentController(targets []corev1.Pod) *PodController {
```
[the remark this answers is not in the document]
Comment (pablochacin) [reply pkg/disruptors/controller.go:135]: >  would expect functions called NewYadaYada to return a YadaYada

Sure, that was the intention. I think it was something I forgot to change when renaming the AgentController to PodControler.  Thanks for the catch.

As for changing  `PodController` to `PodVisitor`, this already exists as an interface. Maybe `PodAgentVisitor`?

File: pkg/disruptors/controller.go [hunk 6]
```diff
@@ -14,40 +14,58 @@ import (
...
 
-// AgentController defines the interface for controlling agents in a set of targets
-type AgentController interface {
-	// InjectDisruptorAgent injects the Disruptor agent in the target pods
-	InjectDisruptorAgent(ctx context.Context) error
-	// Targets retrieves the names of the target of the controller
-	Targets(ctx context.Context) ([]string, error)
-	// Visit allows executing a different command on each target returned by a visiting function
-	Visit(ctx context.Context, visitor PodVisitor) error
+// PodVisitor defines the methods for executing actions in a target Pod
```
[the remark this answers is not in the document]
Comment (pablochacin) [reply pkg/disruptors/controller.go:34]: I don't think the description you suggest is correct `PodVisitor` is an interface. it does not define an object. Objects implement interfaces.

File: pkg/disruptors/pod_test.go (hunk 3 above)
Comment (pablochacin) [reply pkg/disruptors/pod_test.go:23]: There are leftovers from previous versions of the test. Thanks for the catch.

File: pkg/disruptors/controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (pablochacin) [reply pkg/disruptors/controller.go:34]: I don't think this suggestion is correct as `PodVisitor` is an interface, not an object.

File: pkg/disruptors/controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pablochacin) [reply pkg/disruptors/controller.go]: I reverted the change.

**Review by roobre — APPROVED** [review]
Looking awesome to me!

**Pull request merged**